### PR TITLE
Fetch meta files as dill or pickle

### PR DESCRIPTION
**Pull request by ascillitoe**
Fix for #349.

`fetch_state_dict` looks for `dill` files, and then tries `pickle` files. This can be updated to `dill` only in a future PR, once all google bucket artefacts have been updated to `dill`.

### CODE REVIEW

**Comment by jklaise** [comment]
This also seems to address #315.

**Comment by ascillitoe** [comment]
@jklaise yep! Just need to add in some logic for test time. That being said we're only testing `test_fetch_nab` for now.

File: alibi_detect/utils/fetching.py [hunk 1]
```diff
@@ -404,14 +406,36 @@ def fetch_state_dict(url: str, filepath: Union[str, os.PathLike],
...
+        resp.raise_for_status()
+        suffix = '.dill'
+    except RequestException:
+        try:
+            url_meta = os.path.join(url, 'meta.pickle')
+            resp = requests.get(url_meta, timeout=2)
+            resp.raise_for_status()
+            suffix = '.pickle'
+        except RequestException:
+            logger.exception('Timed out which searching for meta.dill or meta.pickle files at {}.'.format(url))
```
Comment (jklaise) [inline alibi_detect/utils/fetching.py:422]: which --> while

I was going to suggest using f-strings, but then I remembered (found?) this: https://blog.pilosus.org/posts/2020/01/24/python-f-strings-in-logging/ Basically it suggests using %-formatting with passing values as arguments to the `exception` method over f-strings or str.format.

File: alibi_detect/utils/fetching.py [hunk 2]
```diff
@@ -404,14 +406,36 @@ def fetch_state_dict(url: str, filepath: Union[str, os.PathLike],
...
+            raise
+
+    # Load metadata and state_dict
+    meta = dill.load(BytesIO(resp.content))
+    try:
+        url_state = os.path.join(url, meta['name'] + suffix)
+        resp = requests.get(url_state)
+        resp.raise_for_status()
+    except RequestException:
+        logger.exception('Timed out which searching for corresponding state file at {}.'.format(url))
```
Comment (jklaise) [inline alibi_detect/utils/fetching.py:432]: Same as above

File: alibi_detect/tests/test_datasets.py [hunk 3]
```diff
@@ -1,20 +1,128 @@
...
+    if return_X_y:
+        assert isinstance(data, tuple)
+        assert isinstance(data[0], pd.DataFrame) and isinstance(data[1], pd.DataFrame)
+    else:
+        assert isinstance(data, Bunch)
+        assert isinstance(data.data, pd.DataFrame) and isinstance(data.target, pd.DataFrame)
+
+
+# Genome dataset
+# TODO - Genome dataset is large compared to others - do we want to include in regular CI?
```
Comment (jklaise) [inline alibi_detect/tests/test_datasets.py:128]: Probably not, some notebooks are excluded for the same reason.

If we did we would need to use our own CI image with all the data pre-loaded, but then testing the fetching part is a moot point as it's already available locally.

Comment (ascillitoe) [reply alibi_detect/tests/test_datasets.py:128]: Fair point. The current tests only add on ~90 secs (locally) so maybe an OK compromise? We could maybe add slowly ones to non-nightly CI when we address #322?

Comment (ascillitoe) [reply alibi_detect/tests/test_datasets.py:128]: I'll remove TODO.

Comment (jklaise) [reply alibi_detect/tests/test_datasets.py:128]: I'd keep the TODO as it keeps a record of the discussion in the source.

**Pull request merged**